src/main/scala/is/hail/expr/Relational.scala

```diff
@@ -194,6 +159,47 @@ object MatrixIR {
        */
     })
   }
+
+  def chooseColsWithArray(typ: MatrixType): (Type, (MatrixValue, Array[Int]) => MatrixValue) = {
+    val rowType = typ.rvRowType
+    val keepType = TArray(+TInt32())
+    val (rTyp, makeF) = ir.Compile[Long, Long, Long]("row", rowType,
+      "keep", keepType,
+      body = InsertFields(ir.Ref("row"), Seq((MatrixType.entriesIdentifier,
+        ir.ArrayMap(ir.Ref("keep"), "i",
+          ir.ArrayRef(ir.GetField(ir.In(0, rowType), MatrixType.entriesIdentifier),
+            ir.Ref("i")))))))
+    assert(rTyp.isOfType(rowType))
+
+    val keepF = { (mv: MatrixValue, keep: Array[Int]) =>
+      val keepBc = mv.sparkContext.broadcast(keep)
+      mv.copy(colValues = keep.map(mv.colValues),
+        rvd = mv.rvd.mapPartitionsPreservesPartitioning(typ.orvdType) { it =>
```

Comment:
we should update typ to have rvRowType `rTyp` otherwise we might smuggle a missing value into a required type.